src/x/cache/lru_cache.go

```diff
@@ -0,0 +1,530 @@
+func (c *LRU) PutWithTTL(key string, value interface{}, ttl time.Duration) {
+	var expiresAt time.Time
+	if ttl > 0 {
+		expiresAt = c.now().Add(ttl)
+	}
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	_, _ = c.updateCacheEntryWithLock(key, expiresAt, value, nil)
+}
+
+// Get returns the value associated with the key, optionally
+// loading it if it does not exist or has expired.
+// NB(mmihic): We pass the loader as an argument rather than
+// making it a property of the cache to support access specific
+// loading arguments which might not be bundled into the key.
+func (c *LRU) Get(ctx context.Context, key string, loader LoaderFunc) (interface{}, error) {
+	return c.GetWithTTL(ctx, key, func(ctx context.Context, key string) (interface{}, time.Time, error) {
+		val, err := loader(ctx, key)
+		return val, time.Time{}, err
+	})
+}
+
+// GetWithTTL returns the value associated with the key, optionally
+// loading it if it does not exist or has expired, and allowing the
+// loader to return a TTL for the resulting value, overriding the
+// default TTL associated with the cache.
+func (c *LRU) GetWithTTL(ctx context.Context, key string, loader LoaderWithTTLFunc) (interface{}, error) {
+	// Spin until it's either loaded or the load fails.
+	for {
+		value, load, loadingCh, err := c.tryCached(key)
+
+		// There was a cached error, so just return it
+		if err != nil {
+			return nil, err
+		}
+
+		// Someone else is loading the entry, wait for this to complete
+		// (or the context to end) and try to acquire again.
+		if loadingCh != nil {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			case <-loadingCh:
+			}
+			continue
+		}
+
+		// No entry exists and no-one else is trying to load it, so we
+		// should try to do so (outside of the mutex lock).
+		if load {
+			if loader == nil {
+				return nil, ErrEntryNotFound
+			}
+
+			return c.tryLoad(ctx, key, loader)
+		}
+
+		// There is an entry and it's valid, return it.
+		return value, nil
+	}
+}
+
+// has checks whether the cache has the given key. Exists only to support tests.
+func (c *LRU) has(key string, checkExpiry bool) bool {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	entry, exists := c.entries[key]
+
+	if !exists {
+		return false
+	}
+
+	if checkExpiry {
+		return entry.loadingCh != nil || entry.expiresAt.After(c.now())
+	}
+
+	return true
+}
+
+// tryCached returns a value from the cache, or an indication of
+// the caller should do (return an error, load the value, wait for a concurrent
+// load to complete).
+func (c *LRU) tryCached(key string) (interface{}, bool, chan struct{}, error) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	entry, exists := c.entries[key]
+
+	// If a load is already in progress, tell the caller to wait for it to finish.
+	if exists && entry.loadingCh != nil {
+		return nil, false, entry.loadingCh, nil
+	}
+
+	// If the entry exists and has not expired, it's a hit - return it to the caller
+	if exists && entry.expiresAt.After(c.now()) {
```

Comment:
Does this need to check that it's not expired? You'll likely be regenerating the same entry anyway here